components/automatic-grid/script.js

```diff
@@ -0,0 +1,244 @@
+
+        //The array of items that have a set position. If an item has only a row or col set, it won't be added here
+        const itemsWithPositions = items.filter((item) => item.hasAttribute('col') && item.hasAttribute('row'));
+
+        //The array of items without a set position. Items with col or row only will also be added here
+        const itemsWithoutPositions = items.filter((item) => itemsWithPositions.indexOf(item) < 0);
+
+        //We add the items with set position to their corresponding cell
+        itemsWithPositions.forEach((item) => {
+            //Get the corresponding cell for that item
+            const matchingCell = this.getMatchingCell(item.getAttribute('col'), item.getAttribute('row'));
+
+            //Check if there is already an item added to that cell. If there is move the item to the other array. This way we avoid items with duplicating col and row
+            if (!matchingCell || matchingCell.childElementCount > 0) {
+                itemsWithoutPositions.push(item);
+                return;
+            }
+
+            matchingCell.appendChild(item);
+        });
+
+        //Append all of the other items to the first available free cell
+        itemsWithoutPositions.forEach((item) => {
+            this.firstEmptyCell.appendChild(item);
+        });
+
+        //If the grid is draggable here we attach the event listener to the items
+        if (this._draggable) {
+            items.forEach((item) => {
+                item.addEventListener('mousedown', this.dragStart);
+            });
+        }
+    }
+
+    /**
+     * Sets the number of columns. If we have more than the max number, the columns are set to the max number. If we don't have a column attribute set, the default is 6
+     * @returns {number} - the number of columns for the grid
+     */
+    setColumns() {
+        let columns = this.getAttribute('columns') || 6;
+
+        if (columns > maxColumns) columns = maxColumns;
+
+        return columns;
+    }
+
+    /**
+     * Helper function to find the cell that is on the specified row and column
+     * @param {Number} col 
+     * @param {Number} row 
+     * @returns {HTMLElement}
+     */
+    getMatchingCell(col, row) {
+        return this._cells.find((cell) => cell.dataset.col === col && cell.dataset.row === row);
+    }
+
+    /**
+     * The function that triggers when we start to drag an item
+     * @param {MouseEvent} event 
+     */
+    dragStart(event) {
+        //We set the draggedItem to the one we are dragging so it can be used in other functions.
+        this._draggedItem = event.currentTarget;
+
+        //Add a class to make the element have a position absolute and become semi-transparent when dragging
+        this._draggedItem.classList.add('guic-dragged');
+
+        //Sets the position of the dragged element to be the same as the mouse
+        this._draggedItem.style.left = `${event.clientX}px`;
+        this._draggedItem.style.top = `${event.clientY}px`;
```

Comment:
If we click on the middle of the element, is this going to move the element's top left corner to the mouse?
Don't we want to offset the position of the element based on the initial mousedown coordinates?
Same for dragMove.